### PR TITLE
New wifi block

### PR DESCRIPTION
**Pull request by MaxVerevkin**
Relevant issue: #1069

This is a simple block that relies on https://github.com/Eonm/nl80211 to get wifi info. I decided to create a new block instead of modifying `net` since it's already bloated (750 SLOC) and I don't want to break it.

This block has four format keys: SSID, signal strength, interface name and wifi frequency.

### TODO
- [ ] Write docs
- [x] Add `{interface}` format key (for example "wlp3s0")
- [ ] Add new `wifi` and `no-wifi` icons. And maybe add `wifi-(1/2/3)` icons for different signal strength.
- [ ] Figure out which runtime dependencies it requires (if any)
- [x] Maybe change color (state) when disconnected?
- [ ] Add an option to hide when disconnected?
- [ ] Deprecate wifi-related options of `net` block and remove them in a later release?

### CODE REVIEW

**Comment by MaxVerevkin** [comment]
For some reason [nl80211](https://github.com/Eonm/nl80211)'s `interface` has a trailing "\u0000". And for some reason `sway`'s parser fails to parse it :eyes:

**Comment by MaxVerevkin** [comment]
Now I wonder if it's possible to disable wifi icon only for `net` block.

**Comment by MaxVerevkin** [comment]
> Now I wonder if it's possible to disable wifi icon only for net block.

I'll just add a new icon "wifi" for the new block.

File: src/widgets/text.rs [hunk 1]
```diff
@@ -69,9 +69,11 @@ impl TextWidget {
...
 
     pub fn set_text(&mut self, content: String) {
-        if content.is_empty() {
-            self.spacing = Spacing::Hidden;
-        }
+        self.spacing = if content.is_empty() {
+            Spacing::Hidden
+        } else {
+            Spacing::Normal
+        };
```
Comment (MaxVerevkin) [inline src/widgets/text.rs:76]: If we already do a check if text is empty, why not check if it's not empty?

It doesn't break any of the blocks I use.

Comment (ammgws) [reply src/widgets/text.rs:76]: Not sure if it is something block author's would actually use, but the only thing I can think of is if a block sets the spacing to something other than `Normal` while still displaying text - this change would change the spacing behind the block's back.

**Comment by ammgws** [comment]
> I decided to create a new block instead of modifying net since it's already bloated (750 SLOC) and I don't want to break it.

To be honest, I would rather this was implemented within the net block (although I understand it will need much more work). From a quick glance, the "bloat" looks like it would mostly go away since a lot of the code could be removed (e.g. the code for parsing SSIDs which is around 250 SLOC by itself.)

- [x] signal_strength: nl80211
- [ ] bitrate: should be possible using netlink
- [ ] ip: should be possible using netlink
- [ ] ipv6: should be possible using netlink
- [x] speed_up: read from /sys/class/net/
- [x] speed_down: read from /sys/class/net/

File: src/widgets/text.rs (hunk 1 above)
Comment (MaxVerevkin) [reply src/widgets/text.rs:76]: Without this change, this config
```toml
[[block]]
block = "net"
device = "..."
format = "{speed_down;K}{speed_up;K}"
format_alt = ""
```
after two clicks has incorrect spacing (no space on the right).

> change the spacing behind the block's back.

It already does it :eyes:

**Comment by MaxVerevkin** [comment]
> bitrate: should be possible using netlink

Does anyone know how?

**Comment by MaxVerevkin** [comment]
I think there are reasons to split `net` and `wifi`. For example, my laptop can be connected to wifi and ethernet at the same time, and I want to know the speed of my active connection (ethernet) and the signal strength (wifi) at the same time. Currently it is not possible. However I understand your point to minimize the number of blocks.

**Comment by ammgws** [comment]
> I want to know the speed of my active connection (ethernet) and the signal strength (wifi) at the same time.

Can't you just have two `net` blocks but with `device` set differently for each one?

**Comment by ammgws** [comment]
No, but that case seems a bit contrived to me. Then again I'm 97% a wired desktop user so perhaps I'm missing out something?

**Comment by MaxVerevkin** [comment]
> but that case seems a bit contrived to me

Fair enough

**Comment by ammgws** [comment]
Sorry I meant the double wifi card thing.

**Comment by MaxVerevkin** [comment]
> Sorry I meant the double wifi card thing.

No, it was just a hypothetical situation which I probably will never face.

**Pull request closed without merging**